File: haskell/glue/src/Glue/Lib/List/Head.hs
module Glue.Lib.List.Head where

import Glue.Eval (Eval, throwError)
import Glue.Eval.Exception (wrongArgumentType)
import Glue.IR (IR (..))

head :: IR Eval
head = NativeFunc headImpl

headImpl :: IR Eval -> Eval (IR Eval)
headImpl arg = case arg of
    List (x : _) -> pure x
    List [] -> throwError $ wrongArgumentType ["non-empty list"]
    _ -> throwError $ wrongArgumentType ["list"]

File: dart/glue/lib/src/lib/list/head.dart
import 'package:glue/src/eval.dart';
import 'package:glue/src/eval/exception.dart';
import 'package:glue/src/ir.dart';

/// Head function - returns the first element of a list
/// Mirrors Haskell Glue.Lib.List.Head.head exactly
Ir head = IrNativeFunc(headImpl);

/// Head function implementation
/// Mirrors Haskell Glue.Lib.List.Head.headImpl exactly
Eval<Ir> headImpl(Ir arg) {
  return switch (arg) {
    IrList(:final elements) =>
      elements.isEmpty
          ? throwError(wrongArgumentType(['non-empty list']))
          : Eval.pure(elements.first),
    _ => throwError(wrongArgumentType(['list'])),
  };
}
